### PR TITLE
ocamlPackages.bls12-381

### PR DESCRIPTION
**Pull request by ulrikstrid**
###### Motivation for this change

This is a continuation of PR https://github.com/NixOS/nixpkgs/pull/129444 which are prerequisites to make LIGO build under nix.
The versions are the specific versions that LIGO needs that's why they are not the latest on opam.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Relase notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by ulrikstrid** [comment]
@vbgl This PR also uses `fetchFromGitLab` now

File: pkgs/development/ocaml-modules/tezos-rust-libs/default.nix [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+    owner = "tezos";
+    repo = "tezos-rust-libs";
+    rev = "v${version}";
+    sha256 = "1ffkzbvb0ls4wk9205g3xh2c26cmwnl68x43gh6dm9z4xsic94v5";
+  };
+
+  buildInputs = [
+    topkg
+    findlib
+  ];
```
Comment (vbgl) [inline pkgs/development/ocaml-modules/tezos-rust-libs/default.nix:17]: Don’t you need `ocaml` there? Why did you add `topkg` and `findlib`?

File: pkgs/development/ocaml-modules/tezos-rust-libs/default.nix [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+  propagatedBuildInputs = [ cargo rustc ];
```
Comment (vbgl) [inline pkgs/development/ocaml-modules/tezos-rust-libs/default.nix:19]: This should probably not be propagated.

File: pkgs/development/ocaml-modules/tezos-rust-libs/default.nix [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+  doCheck = true;
+
+  meta = {
+    homepage = "https://gitlab.com/tezos/tezos-rust-libs";
+    description = "Tezos: all rust dependencies and their dependencies";
+    license = lib.licenses.mit;
+    maintainers = [ lib.maintainers.ulrikstrid ];
+  };
+
+  createFindlibDestdir = true;
```
Comment (vbgl) [inline pkgs/development/ocaml-modules/tezos-rust-libs/default.nix:40]: Is this needed?

File: pkgs/development/ocaml-modules/bls12-381/default.nix [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+    owner = "dannywillems";
+    repo = "ocaml-bls12-381";
+    rev = version;
+    sha256 = "1s8n657fsl2gs01p7v2ffpcfzymavifhhpriyx1gq5qh4zvvw4vr";
+  };
+  useDune2 = true;
+
+  propagatedBuildInputs = [
+    ff
+    dune-configurator
```
Comment (vbgl) [inline pkgs/development/ocaml-modules/bls12-381/default.nix:17]: Is `dune-configurator` needed? It should not be propagated.

**Review by vbgl — COMMENTED** [review]
Does `tezos-rust-libs` belong to `ocamlPackages`? I.e., does the version of OCaml used to build it have an impact on the result?

File: pkgs/development/ocaml-modules/bls12-381/default.nix (hunk 4 above)
Comment (ulrikstrid) [reply pkgs/development/ocaml-modules/bls12-381/default.nix:17]: Seems like it's not needed actually, removing it

File: pkgs/development/ocaml-modules/tezos-rust-libs/default.nix (hunk 3 above)
Comment (ulrikstrid) [reply pkgs/development/ocaml-modules/tezos-rust-libs/default.nix:40]: It seems like this line is not needed, but the install fails if I don't have findlib as a dependency

File: pkgs/development/ocaml-modules/tezos-rust-libs/default.nix (hunk 2 above)
Comment (ulrikstrid) [reply pkgs/development/ocaml-modules/tezos-rust-libs/default.nix:19]: Correct, thank you

File: pkgs/development/ocaml-modules/tezos-rust-libs/default.nix (hunk 1 above)
Comment (ulrikstrid) [reply pkgs/development/ocaml-modules/tezos-rust-libs/default.nix:17]: the install fails if I don't have `findlib` as a dependency, I guess it's because I'm using `$OCAMLFIND_DESTDIR` in the `installPhase`

Comment (ulrikstrid) [reply pkgs/development/ocaml-modules/tezos-rust-libs/default.nix:17]: `topkg` can be removed, thank you

**Comment by ulrikstrid** [comment]
> Does `tezos-rust-libs` belong to `ocamlPackages`? I.e., does the version of OCaml used to build it have an impact on the result?
It doesn't use OCaml directly but it is a opam library, I'm not sure what to do to make this better?

**Comment by vbgl** [comment]
I tried to build tezos-rust-lib with different versions of the OCaml compiler, and the results were identical. I suggest to move this out of the `ocamlPackages` set and to install it in a more usual path (`$out/lib`).

**Comment by ulrikstrid** [comment]
Will this still work then? https://github.com/NixOS/nixpkgs/pull/129509/files#diff-7c5a8b3035e1cea4a96396db202c7618d9e2b7f07f17620b523aa0a00db67f04R27

**Comment by vbgl** [comment]
You can simplify it to `OPAM_SWITCH_PREFIX = "${tezos-rust-libs}";`.

**Comment by ulrikstrid** [comment]
Moved it into `libraries` now, had to do some extra copies to get everything in place but it should be fine

File: pkgs/development/ocaml-modules/bls12-381/default.nix [hunk 5]
```diff
@@ -0,0 +1,37 @@
+{ lib, fetchFromGitLab, buildDunePackage, ff, zarith, ctypes, tezos-rust-libs, ocaml, alcotest }:
```
Comment (vbgl) [inline pkgs/development/ocaml-modules/bls12-381/default.nix:1]:
```suggestion
{ lib, fetchFromGitLab, buildDunePackage, ff, zarith, ctypes, tezos-rust-libs, alcotest }:
```

File: pkgs/development/ocaml-modules/bls12-381/default.nix [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+  version = "0.3.15";
+
+  src = fetchFromGitLab {
+    owner = "dannywillems";
+    repo = "ocaml-bls12-381";
+    rev = version;
+    sha256 = "1s8n657fsl2gs01p7v2ffpcfzymavifhhpriyx1gq5qh4zvvw4vr";
+  };
+  useDune2 = true;
+
```
Comment (vbgl) [inline pkgs/development/ocaml-modules/bls12-381/default.nix:14]:
```suggestion

  minimalOCamlVersion = "4.08";

```
